packages/client/lib/client/parser.ts

```diff
@@ -0,0 +1,122 @@
+import { RedisArgument, RespVersions } from "../RESP/types";
+import { RedisVariadicArgument } from "../commands/generic-transformers";
+
+export interface CommandParser {
+  redisArgs: ReadonlyArray<RedisArgument>;
+  keys: ReadonlyArray<RedisArgument>;
+  firstKey: RedisArgument | undefined;
+  respVersion: RespVersions;
+  preserve: unknown;
+  cachable: boolean;
+
+  push: (...arg: Array<RedisArgument>) => unknown;
+  pushVariadic: (vals: RedisVariadicArgument) => unknown;
+  pushVariadicWithLength: (vals: RedisVariadicArgument) => unknown;
+  pushVariadicNumber: (vals: number | Array<number>) => unknown;
+  pushKey: (key: RedisArgument) => unknown; // normal push of keys
+  pushKeys: (keys: RedisVariadicArgument) => unknown; // push multiple keys at a time
+  pushKeysLength: (keys: RedisVariadicArgument) => unknown; // push multiple keys at a time
+  setCachable: () => unknown;
+  setPreserve: (val: unknown) => unknown;
+}
+
+export class BasicCommandParser implements CommandParser {
+  #redisArgs: Array<RedisArgument> = [];
+  #keys: Array<RedisArgument> = [];
+  #respVersion: RespVersions;
+  #preserve: unknown;
```

Comment:
remove the private nature, and get rid of getter and setPreserve and access directly.